A43-sdk-authorization.md

```diff
@@ -0,0 +1,713 @@
+        .addService(service)
+        .intercept(authzServerInterceptor)
+        .build()
+        .start();
+```
+
+- Policy is reloaded dynamically from file path
+
+```Java
+ScheduledExecutorService scheduledExecutor =
+    Executors.newSingleThreadScheduledExecutor(
+        new ThreadFactoryBuilder()
+            .setNameFormat(...)
+            .setDaemon(true)
+            .build());
+FileAuthorizationServerInterceptor authzServerInterceptor;
+try {
+  authzServerInterceptor = FileAuthorizationServerInterceptor.create(authzPolicyFile);
+} catch (IllegalArgumentException iae) {
+  // ...
+} catch (IOException ioe) {
+  // ...
+}
+Closeable closeable = authzServerInterceptor.scheduleRefreshes(
+    1, TimeUnit.HOURS, scheduledExecutor);
+Server server =
+    Grpc.newServerBuilderForPort(port, serverCreds)
+        .addService(service)
+        .intercept(authzServerInterceptor)
+        .build()
+        .start();
+// In the end, free up the resources used for policy refresh.
+closeable.close();
+```
+
+## Rationale
+
+gRPC Authorization internally implements RBAC Engine(s) based on Envoy [RBAC policy](https://github.com/envoyproxy/envoy/blob/main/api/envoy/config/rbac/v3/rbac.proto).
+We decided to create a new policy language "SDK authorization policy" instead of 
+consuming Envoy RBAC directly due to following reasons:
+- Envoy RBAC is a complex language, and we preferred using a simple human readable 
+  policy language
+- With our own language, we can provide a stable API, even when Envoy undergoes 
+  versioning updates.
+
+Note that SDK authorization policy is a subset of Envoy RBAC, and it does not
+support all the fields that are present in Envoy RBAC. 
+
+RBAC policy provides service-level and method-level access control for a service.
+Engines process incoming RPC request attributes against policy configs and make a
+decision on whether to allow or deny the request. The decision depends on the type
+of policy (if the policy is an allowlist or denylist) and whether a matching policy
+was found.
+
+This engine implementation is shared by SDK-based authorization and xDS-based
+authorization. In gRPC Proxyless Service Mesh (or xDS authorization), Istio or
+Traffic Director control plane sends Envoy RBAC policy configs to xDS enabled gRPC
+servers for authorization. ([A41](https://github.com/grpc/proposal/pull/237): xDS RBAC Support)
+
+Overall SDK authorization flow is as follows. User supplies gRPC SDK authorization
+policy to provider/interceptor. The provider then forwards the JSON policy to Policy
+translator. The translator converts JSON policy to Envoy RBAC protos (Allow and/or
+Deny policy). Translator errors out on I/O errors or if the policy does not
+represent JSON schema it currently supports. Ultimately the generated RBAC policies
+are used to create Envoy RBAC authorization engine(s). Then, for each incoming RPC
+request, we will invoke the Engines to get the authorization decision.
```

Comment:
Any mention of optimization/caching such as:

- when a policy contains only principals (and no paths/headers) it's result is the same for all RPCs on a connection since the principals for a connection don't change
- if you cache the authorization result for a given set of headers+paths+principals then the cache can be used instead of evaluating the policy for every single RPC. 

Are they just implementation details or are they not being considered for now?

Reply:
Currently we do not have any such optimization/caching. For each incoming RPC, we evaluate against the SDK policy. I think several things need to be considered when doing such optimizations, especially with file reloading support. Being aware of the memory usage. Developing an efficient caching structure that can scale as the policy grows. Having some sort of monitoring to actually see the optimization benefits etc.
Thanks for bringing this up. We can definitely pick this up later.